File: binary-subarrays-with-sum/binary-subarrays-with-sum.cpp

class Solution {
public:
    int numSubarraysWithSum(vector<int>& A, int S) {
        int sum = 0, start = 0,ans = 0;
        for(int i = 0 ; i < A.size() ; i++){
            sum+=A[i];
            if(sum > S){
                for( ; start < i && sum > S ; ){
                    sum -=A[start];
                    start++;
                }       
            }
            if(sum == S){
                ans+=1;
                for(int x = start ; x < i && sum-A[x] == S ; x++){
                    ans+=1;
                }
            }
        }
        return ans;
    }
};

File: binary-subarrays-with-sum/binary-subarrays-with-sum.go

func numSubarraysWithSum(A []int, S int) int {
    ans , start,sum := 0,0,0
    for i := 0 ; i < len(A) ; i++{
        sum+=A[i]
        if sum > S{
            for ; sum > S && start < i ; {
                sum -= A[start]
                start+=1
                
            }
        }
         if sum == S {
            ans+=1
             x := start
             for ; x < i && sum - A[x] == S ; {
                 x+=1
                 ans+=1
             }
          
        }
    }
    return ans
    
}
